thoth/python/aiosource.py

```diff
@@ -0,0 +1,414 @@
+#!/usr/bin/env python3
+# thoth-python
+# Copyright(C) 2018, 2019 Fridolin Pokorny
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Representation of source (index) for Python packages."""
+
+import logging
+import re
+import typing
+from functools import lru_cache
+from urllib.parse import urlparse
+
+import attr
+import aiohttp
+
+from bs4 import BeautifulSoup
+
+import semantic_version as semver
+
+from .exceptions import NotFound
+from .exceptions import InternalError
+from .exceptions import VersionIdentifierError
+from .configuration import config
+from .artifact import Artifact
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@attr.s(frozen=True, slots=True)
+class AIOSource:
```

Comment:
Will it be possible to derive from the old `Source` class and implement additional async methods not to copy the implementation of the existing methods?

Reply:
cant we refactor all the logic into a common base class and just have some sync and some async `_do_request()`?

Reply:
+1 on my side.